src/Discord/Internal/Types/Interactions.hs

```diff
@@ -0,0 +1,365 @@
+              ("options", toJSON <$> applicationCommandInteractionDataOptionOptions),
+              ("focused", toJSON <$> applicationCommandInteractionDataOptionFocused)
+            ]
+      ]
+
+instance FromJSON ApplicationCommandInteractionDataOption where
+  parseJSON =
+    withObject
+      "ApplicationCommandInteractionDataOption"
+      ( \v ->
+          ApplicationCommandInteractionDataOption
+            <$> v .: "name"
+            <*> v .: "type"
+            <*> v .:? "value"
+            <*> v .:? "options"
+            <*> v .:? "focused"
+      )
+
+-- | The data to respond to an interaction with. Unless specified otherwise, you
+-- only have three seconds to reply to an interaction before a failure state is
+-- given.
+data InteractionResponse = InteractionResponse
+  { interactionResponseType :: InteractionCallbackType,
+    interactionResponseData :: Maybe InteractionCallbackData
+  }
+  deriving (Show, Read, Eq)
+
+instance Default InteractionResponse where
+  def = InteractionResponse InteractionCallbackTypeChannelMessageWithSource Nothing
+
+instance ToJSON InteractionResponse where
+  toJSON InteractionResponse {..} =
+    object
+      [ (name, value)
+        | (name, Just value) <-
+            [ ("type", toMaybeJSON interactionResponseType),
+              ("data", toJSON <$> interactionResponseData)
+            ]
+      ]
+
+-- | What's the type of the response?
+data InteractionCallbackType
+  = -- | Responds to a PING.
+    InteractionCallbackTypePong
+  | -- | Respond with a message to the interaction
+    InteractionCallbackTypeChannelMessageWithSource
+  | -- | Respond with a message to the interaction, after a delay. Sending this
+    -- back means the interaction token lasts for 15 minutes.
+    InteractionCallbackTypeDeferredChannelMessageWithSource
+  | -- | For components, edit the original message later.
+    InteractionCallbackTypeDeferredUpdateMessage
+  | -- | For components, edit the original message.
+    InteractionCallbackTypeUpdateMessage
+  | -- | Respond to an autocomplete interaction with suggested choices.
+    InteractionCallbackTypeApplicationCommandAutocompleteResult
+  deriving (Show, Read, Eq, Data)
+
+instance Enum InteractionCallbackType where
+  fromEnum InteractionCallbackTypePong = 1
+  fromEnum InteractionCallbackTypeChannelMessageWithSource = 4
+  fromEnum InteractionCallbackTypeDeferredChannelMessageWithSource = 5
+  fromEnum InteractionCallbackTypeDeferredUpdateMessage = 6
+  fromEnum InteractionCallbackTypeUpdateMessage = 6
+  fromEnum InteractionCallbackTypeApplicationCommandAutocompleteResult = 6
+  toEnum a = fromJust $ lookup a table
+    where
+      table = makeTable InteractionCallbackTypePong
+
+instance ToJSON InteractionCallbackType where
+  toJSON = toJSON . fromEnum
+
+-- | Convenience wrapper for two separate types of callback.
+data InteractionCallbackData = ICDM InteractionCallbackMessages | ICDA InteractionCallbackAutocomplete
+  deriving (Show, Read, Eq)
```

Comment:
`ICDM` and `ICDA` are weird. 

```haskell
data InteractionCallbackData =  InteractionCallbackDataMessage InteractionCallbackMessages 
                             | InteractionCallbackDataAutocomplete InteractionCallbackAutocomplete
```

Long unwieldy, but explicit. It's uncommon enough that I prefer to follow the ugly rule than create a short exception